src/py-opentimelineio/opentimelineio/adapters/otiod.py

```diff
@@ -0,0 +1,146 @@
+a single directory named with a suffix of .otiod.
+"""
+
+import os
+import copy
+import shutil
+
+from . import (
+    file_bundle_utils as utils,
+    otio_json,
+)
+
+from .. import (
+    exceptions,
+    schema,
+)
+
+
+def read_from_file(filepath, absolute_media_reference_paths=False):
+    result = otio_json.read_from_file(
+        os.path.join(filepath, utils.BUNDLE_PLAYLIST_PATH)
+    )
+
+    if not absolute_media_reference_paths:
+        return result
+
+    for cl in result.each_clip():
+        try:
+            source_fpath = cl.media_reference.target_url
+        except AttributeError:
+            continue
+
+        new_fpath = "file://{}".format(
+            os.path.join(
+                filepath,
+                source_fpath.split("file://")[1]
+            )
+        )
+
+        cl.media_reference.target_url = new_fpath
+
+    return result
+
+
+def write_to_file(
+    input_otio,
+    filepath,
+    media_policy=utils.MediaReferencePolicy.ErrorIfNotFile,
+    dryrun=False
+):
+    # make sure the incoming OTIO isn't edited
+    input_otio = copy.deepcopy(input_otio)
+
+    manifest = utils._file_bundle_manifest(
+        input_otio,
+        filepath,
+        media_policy,
+        "OTIOD"
+    )
+
+    # dryrun reports the total size of files
+    if dryrun:
+        fsize = 0
+        for fn in manifest:
+            fsize += os.path.getsize(fn)
+        return fsize
+
+    fmapping = {}
+
+    # gather the files up in the staging_dir
+    for fn in manifest:
+        target = os.path.join(
+            filepath,
+            utils.BUNDLE_DIR_NAME,
+            os.path.basename(fn)
+        )
+        fmapping[fn] = target
+
+    # so we don't edit the incoming file
+    input_otio = copy.deepcopy(input_otio)
+
+    # update the media reference
+    for cl in input_otio.each_clip():
+        if media_policy == utils.MediaReferencePolicy.AllMissing:
+            cl.media_reference = schema.MissingReference()
+            continue
+
+        try:
+            source_fpath = cl.media_reference.target_url
+        except AttributeError:
+            continue
+
+        basename = os.path.basename(source_fpath.split("file://")[1])
+        newpath = os.path.join("media", basename)
+
+        cl.media_reference.target_url = "file://{}".format(newpath)
+
+    if not os.path.exists(os.path.dirname(filepath)):
+        raise exceptions.OTIOError(
+            "Error: directory '{}' does not exist, cannot create '{}'".format(
+                os.path.dirname(filepath),
+                filepath
+            )
+        )
+    os.mkdir(filepath)
+
+    # write the otioz file to the temp directory
+    otio_json.write_to_file(
+        input_otio,
+        os.path.join(filepath, utils.BUNDLE_PLAYLIST_PATH)
+    )
+
+    # write the media files
+    os.mkdir(os.path.join(filepath, utils.BUNDLE_DIR_NAME))
+    for src, dst in fmapping.items():
+        shutil.copyfile(src, dst)
```

Comment:
Would it be an idea to abstract the file copy into a function that attempts to use hard links on supported OS and when files exist on the same server? It would speed up the process and save disk space.

Reply:
Could quite simply be done like this as well:

``` python
try:
    os.link(src, dst)

except (OSError, AttributeError):
    # OSError - Perhaps file resides on another storage device
    # AttributeError - Windows does not support hard links and thus don't supply os.link()
    shutil.copyfile(src, dst)

```

Reply:
I like the idea but I'd want to make this an enhancement on the basic thing rather than do this out of the gate.  Lets keep it simple for now and see what kinds of options people need/want.  What you're proposing should be straightforward to add in a later PR.  Thanks!